### PR TITLE
Cask core code moved to Homebrew

### PR DESCRIPTION
**Pull request by AnastasiaSulyagina**
### Changes to the core
- [x] Followed [hacking.md](https://github.com/caskroom/homebrew-cask/blob/master/doc/development/hacking.md).
## 

Moving [Cask](https://github.com/caskroom/homebrew-cask) core code to Homebrew as part of [#14384](https://github.com/caskroom/homebrew-cask/issues/14384).

**Do not merge before [#725](https://github.com/Homebrew/brew/pull/725) is merged.** Will rebase after everything is reviewed.

cc @MikeMcQuaid @vitorgalvao @jawshooah @adidalal (bring other cask maintainers please)

### CODE REVIEW

**Comment by jawshooah** [comment]
Pinging @caskroom/maintainers for visibility.

**Comment by MikeMcQuaid** [comment]
Unfortunately there's already conflicts that need resolved although I guess that could in theory be done by whoever is merging. 👍 from the Homebrew side to doing this if all the tests are passing and a `brew cask install` works etc.

**Comment by AnastasiaSulyagina** [comment]
There will be some conflicts by the time we are merging it anyway so I would better resolve all at once in the end when have LGTM.

**Comment by MikeMcQuaid** [comment]
I'm 👍 to do this after https://github.com/Homebrew/brew/pull/725 is green (probably after https://github.com/Homebrew/brew/pull/727 is merged).

Adminy things for Cask maintainers:
- We need to make sure to merge https://github.com/Homebrew/brew/pull/725 first and then this PR shortly afterwards. It may be worth syncing these two operations on e.g. Slack or similar
- Let me know who needs commit access to Homebrew/brew. I wouldn't imagine every Cask maintainer past and present will need commit access. Read through https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/New-Maintainer-Checklist.md and make sure there's nothing you find problematic. We'll have a polite agreement for Cask maintainers to not merge changes to other parts of Homebrew/brew and vice-versa until the project boundaries have blurred a bit more.
- Let me know ASAP any other concerns/complaints you have and I'll work ASAP to address them

**Comment by MikeMcQuaid** [comment]
In case other maintainers are interested to click around and see what the Cask repo will look like after merging: see https://github.com/caskroom/homebrew-cask/tree/24d9e370b8eb6d63791f9cf235c1a11091ac7d63

**Comment by AnastasiaSulyagina** [comment]
And I probably would need some advice about documentation split: I did not figure out what should stay and what should be moved.

**Comment by jawshooah** [comment]
@AnastasiaSulyagina You'll also want to include the core-specific dotfiles: `.rspec`, `.rubocop.yml`, and `.simplecov`.

**Comment by adidalal** [comment]
Some notes on my future involvement:

I don't particularly touch core code and don't see any need for me to need commit access to Homebrew/brew.

There are a few aspects of Caskroom documentation that I really like, and would like to bring over to the brew docs, but that's a discussion for another PR.

Going forward if homebrew/binary is going away, I'd like to be a part of that transition of Formulas to Casks.

**Comment by MikeMcQuaid** [comment]
> Going forward if homebrew/binary is going away, I'd like to be a part of that transition of Formulas to Casks.

It is, yep. Feel free to just start creating PRs to add [these](https://github.com/homebrew/homebrew-binary) to Cask and we can remove them from homebrew/binary after that 👍

File: ci/travis/script.sh [hunk 1]
```diff
@@ -18,6 +18,5 @@ fi
 
 if must_run_tests; then
   run bundle exec rake rubocop
-  run brew cask-tests --coverage
```
Comment (jawshooah) [inline ci/travis/script.sh]: If you're just trying to get Travis to go green, you may as well remove this entire `if` block, along with anything else in these scripts related to testing the core.

Comment (MikeMcQuaid) [inline ci/travis/script.sh]: @jawshooah 🆒, thanks. This was my suggestion as I wasn't sure what else here was or wasn't just for core testing.

**Comment by AnastasiaSulyagina** [comment]
Build fails because brew PR is not merged. Maybe it would be reasonable to first merge #725 and then fix the build and everything. On my machine everything works when cask is in both brew and here. What do you think @jawshooah?

**Comment by jawshooah** [comment]
Yup, should definitely merge the brew PR first, otherwise we'll break everything for users :bomb:

**Comment by AnastasiaSulyagina** [comment]
With 'better' I meant 'better than fixing travis in some magic ways', not 'better than merging this first' :D

**Comment by MikeMcQuaid** [comment]
Have merged https://github.com/Homebrew/brew/pull/725 and invited @jawshooah @vitorgalvao @reitermarkus to the Homebrew/brew repository, @Homebrew/cask team and @Homebrew organisation.

**Comment by MikeMcQuaid** [comment]
Post https://github.com/Homebrew/brew/pull/725 being merged it's still prioritising the tap `brew cask` over the new code in Homebrew/brew. My personal testing (by commenting out the code adding the tap code to the PATH) shows that both new and old `brew cask`s are still working.

**Comment by jasonkarns** [comment]
`brew doctor` is now failing presumably for people who have the old tap and now have cask in core?

```
$ brew doctor 
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: You have external commands with conflicting names.

Found command `brew-cask-tests` in following places:
    /usr/local/Library/Taps/caskroom/homebrew-cask/cmd/brew-cask-tests.rb
    /usr/local/Library/Homebrew/cask/cmd/brew-cask-tests.rb

Found command `brew-cask` in following places:
    /usr/local/Library/Taps/caskroom/homebrew-cask/cmd/brew-cask.rb
    /usr/local/Library/Homebrew/cask/cmd/brew-cask.rb
```

Issue is resolved by `brew untap caskroom/cask`. Sharing this for google-ability. Maybe there should be some messaging or warning about this for existing cask users?

**Comment by MikeMcQuaid** [comment]
@jasonkarns Gonna remove that doctor warning ASAP although there is a note saying to ignore these if there's no issues 😉

**Comment by MikeMcQuaid** [comment]
Done in https://github.com/Homebrew/brew/pull/752

**Comment by MikeMcQuaid** [comment]
@jawshooah If you get a sec could you trigger the rebuild of these Travis CI tests? Thanks!

**Comment by reitermarkus** [comment]
Merged! 🎉

**Comment by MikeMcQuaid** [comment]
@AnastasiaSulyagina 👏  and @reitermarkus thanks for the merge!

**Pull request merged**